[cuda] Move to hal/drivers and wire up BUILD files

antiagainst opened the pull request:
This commit moves the CUDA HAL driver rewrite to the
`hal/drivers` directory given it's functional and ready for
normal usage. By this we can start run tests with CI
to make sure it does not regress. Further improvements
can happen directly in this directory.

This provides an easy route for trying out the rewrite before
eventually replace the existing HAL driver. 

Along the way wired up BUILD configurations.

Progress towards https://github.com/openxla/iree/issues/13245

benvanik commented on runtime/src/iree/hal/drivers/cuda2/cuda_device.h:
```diff
@@ -7,11 +7,11 @@
 #ifndef EXPERIMENTAL_CUDA2_CUDA_DEVICE_H_
 #define EXPERIMENTAL_CUDA2_CUDA_DEVICE_H_
```
header guards need updating

ScottTodd commented on CMakeLists.txt:
```diff
@@ -226,6 +226,7 @@ option(IREE_HAL_DRIVER_DEFAULTS "Sets the default value for all runtime HAL driv
 set(IREE_HAL_DRIVER_CUDA_DEFAULT ${IREE_HAL_DRIVER_DEFAULTS})
 if(NOT IREE_CUDA_AVAILABLE OR CMAKE_CROSSCOMPILING)
   set(IREE_HAL_DRIVER_CUDA_DEFAULT OFF)
+  set(IREE_HAL_DRIVER_CUDA2_DEFAULT OFF)
```
Initialize to `ON` if available? Should be explicit either way.
```suggestion
set(IREE_HAL_DRIVER_CUDA_DEFAULT ${IREE_HAL_DRIVER_DEFAULTS})
set(IREE_HAL_DRIVER_CUDA2_DEFAULT ${IREE_HAL_DRIVER_DEFAULTS})
if(NOT IREE_CUDA_AVAILABLE OR CMAKE_CROSSCOMPILING)
  set(IREE_HAL_DRIVER_CUDA_DEFAULT OFF)
  set(IREE_HAL_DRIVER_CUDA2_DEFAULT OFF)
```

antiagainst replied:
Good catch. Missed that. Fixed.

ScottTodd commented on runtime/src/iree/hal/drivers/cuda2/README.md:
Could tweak the README while moving it. In particular this section: https://github.com/openxla/iree/blob/4a347236e7a9ab38b4dffe9e089e72e15cf3822c/experimental/cuda2/README.md?plain=1#L6-L10

(also note that the issue link there does not render on GitHub, see https://github.com/openxla/iree/blob/main/experimental/cuda2/README.md)

antiagainst replied:
Revised.

ScottTodd commented on runtime/src/iree/hal/drivers/cuda2/cts/CMakeLists.txt:
```diff
@@ -8,15 +8,15 @@ iree_hal_cts_test_suite(
   DEPS
-    iree::experimental::cuda2::registration
+    iree::hal::drivers::cuda2::registration
   EXCLUDED_TESTS
     # HAL event is unimplemented for now.
     "event"
```
Are events still unimplemented?

antiagainst replied:
Right. The compiler side does not generate this at the moment; so it's fine.

ScottTodd commented on CMakeLists.txt:
```diff
@@ -243,6 +245,7 @@ if(NOT APPLE OR NOT ${CMAKE_SYSTEM_PROCESSOR} MATCHES "arm64")
 endif()
 
 option(IREE_HAL_DRIVER_CUDA "Enables the 'cuda' runtime HAL driver" ${IREE_HAL_DRIVER_CUDA_DEFAULT})
+option(IREE_HAL_DRIVER_CUDA2 "Enables the 'cuda2' runtime HAL driver" ${IREE_HAL_DRIVER_CUDA2_DEFAULT})
```
Can we come up with a plan for switching from the old `cuda` to the new `cuda2`? Maybe post an RFC and an update to https://github.com/openxla/iree/issues/13245, and could link to one or both of those here.

Some ideas:

* Drive to feature complete
  * All CTS tests and execution tests we care about pass using `cuda2`
  * Figure out the story for flags like on https://github.com/openxla/iree/issues/13984
* Switch default in-repo
  * All in-repo benchmarks using `cuda2`, without egregious regressions
* Delete (old) `cuda` and rename `cuda2` to `cuda`

Ideally we'd have a timeline in mind so we don't stay in this state of having two similar paths for long.

antiagainst replied:
SG. I've commented on #13245 with https://github.com/openxla/iree/issues/13245#issuecomment-1684210183. I'll send out a mail later too.

antiagainst replied:
Hard for me to give a precise timeline though; given I'm multiplexing over quite a few work streams. But yeah agreed we don't want to be in a hybrid state for too long; I'm hoping to get feature parity down soon.

antiagainst committed:
[cuda] Move to hal/drivers and wire up BUILD files

This commit moves the CUDA HAL driver rewrite to the
`hal/drivers` directory given it's functional and ready for
normal usage. By this we can start run tests with CI
to make sure it does not regress. Further improvements
can happen directly in this directory.

This provides an easy route for trying out the rewrite before
eventually replace the existing HAL driver.

Along the way wired up BUILD configurations.
